universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java

```diff
@@ -125,19 +126,77 @@ private void putNull(Path path) {
    * Puts the given value at the given path in the underlying JSON data. Builds up the necessary
    * structure along the way, i.e., creates parent objects where necessary.
    *
-   * @param path the {@link Path} with the fully specified path, e.g., "applicant.favorites.color"
-   *     or the equivalent "$.applicant.favorite.colors".
+   * <p>For internal lists
+   *
+   * @param path the {@link Path} with the fully specified path, e.g.,
+   *     "applicant.children[3].favorite_color.text" or the equivalent
+   *     "$.applicant.children[3].favorite_color.text".
    * @param value the value to place; values of type Map will create the equivalent JSON structure
    */
   private void put(Path path, Object value) {
-    if (path.parentPath().isEmpty()) {
-      jsonData.put(Path.JSON_PATH_START_TOKEN, path.keyName(), value);
+    maybePutParent(path);
+    putAt(path, value);
+  }
+
+  private void putAt(Path path, Object value) {
+    jsonData.put(path.parentPath().toString(), path.keyName(), value);
+  }
+
+  private void addAt(Path path, Object value) {
+    jsonData.add(path.withoutArrayReference().toString(), value);
+  }
+
+  /**
+   * Put parent of path if it doesn't exist. There are two types of parents: JSON objects and JSON
+   * arrays.
+   *
+   * <p>For JSON object parents, if it doesn't already exist an empty map is put in the right place.
+   *
+   * <p>For JSON array parents, if the array (e.g. applicant.children[]) doesn't already exist an
+   * empty array is put in the right place, and then if the array element (e.g.
+   * applicant.children[3]) doesn't already exist then empty maps are added until an empty map is
+   * available at the right index.
+   */
+  private void maybePutParent(Path path) {
+    Path parentPath = path.parentPath();
+
+    if (hasPath(parentPath)) {
       return;
     }
-    if (!hasPath(path.parentPath())) {
-      put(path.parentPath(), new HashMap<>());
+
+    // TODO(#624): get rid of this recursion.
+    maybePutParent(parentPath);
+
+    if (parentPath.isArrayElement()) {
+      putParentList(path);
+    } else {
+      putAt(parentPath, new HashMap<>());
+    }
+  }
+
+  /**
+   * Put parent of path if it doesn't already exist, for parents that are lists (e.g. something[n]),
```

Comment:
Suggestion (blocking): the terminology for repeated answers uses "list" and "array" interchangeably. Recommend standardizing on "array" since that's the word for it in JSON.

Reply:
done!

Reply:
thanks! still have `putParentList`

Reply:
ahht thanks. i didn't search for List. :mag: